### ISSUE TITLE
[Question] How to get observation of added agent in Parallel API?

### ISSUE THREAD

**Issue by Pyrolistical** [Farama-Foundation/PettingZoo#989]
### Question

According to https://pettingzoo.farama.org/api/parallel/#pettingzoo.utils.env.ParallelEnv.agents, `.agents` is:
> A list of the names of all current agents, typically integers. These may be changed as an environment progresses (i.e. **agents can be added** or removed).

For agents that are added later, how do we get the observation? We need the observation in order to decide what action to take.

Is the `.step()` suppose to return observation for newly added agents? That would mean the `agent_id` keys in `actions` are different than the returns `observations`.

**Comment by elliottower** [comment]
The `step()` function will return information for all agents which are currently alive, the keys will change accordingly (when an agent dies/terminates, their keys are removed from the observation as well as rewards, infos, etc.

For examples to reference: Knights Archers Zombies environment I know adds agents, although it is a bit complicated (source code: https://pettingzoo.farama.org/_modules/pettingzoo/butterfly/knights_archers_zombies/knights_archers_zombies/#raw_env), I believe MPE environments do as well but can't confirm.

**Comment by Pyrolistical** [comment]
The knights archer zombies env is implemented using ACE api which has [`.observe(agent)`](https://pettingzoo.farama.org/api/aec/#pettingzoo.utils.env.AECEnv.observe). 

If the Parallel api also had `.observe(agent)` then this wouldn't be a problem. 

The more I think about it, the more I don't like how the Parallel api tries to be gym like. Feels like there could be a simpler api

**Comment by elliottower** [comment]
Sorry for the late response but you do raise a good point, we'll look into it internally. If you'd like to contribute a PR or explain your reasoning a bit further that would be helpful.

**Comment by elliottower** [comment]
Just circling back on this, do you still think it would be useful to add an `observe()` method to the Parallel API? It does feel simpler to be honest, and you can always just ignore it. Keeping the return values on `env.step()` I think is good though, wouldn't want to remove it and change all the backwards compatibility. It also means gymnasium code like vector envs and functional API can pretty easily be ported into the parallel API because they behave similarly.

**Comment by Pyrolistical** [comment]
I think that is probably the easiest fix if you want to keep the API gym like otherwise

**Comment by elliottower** [comment]
> I think that is probably the easiest fix if you want to keep the API gym like otherwise

Makes sense, will incorporate soon and make a new release in the next week with it

**Comment by elliottower** [comment]
Update: I see there actually isn't an `observation` dict for some reason, I guess they want you to call `observe` because observation can sometimes be stochastic and sample from the environment (you can seed the observation spaces for example, and we have the `from pettingzoo.test.example_envs import generated_agents_env_v0` env which has randomized obs). 

This means that simply storing the previous observation value would not be sufficient to do an `observe` method. The only thing which would achieve this sort of sampling would be to do a deepcopy of the environment at every step, and then you can 'recreate' the last step of the environment repeatedly in order to get new observation values. However, that is not clean at all and would introduce a lot of overhead. 

Environments which were written as `def raw_env(ParallelEnv)` would need to be rewritten in order to have an explicit observe method, but it actually turns out that most of the reference environments in PettingZoo were written as AECEnv and are just converted on the fly using the `aec_to_parallel` function. So it would be possible for me to change that function to also incorporate an `observe` method using the AEC observe methods for each agent. Would be a little bit complicated but would be nice to have IMO. The only issue is that as I mentioned earlier, some envs would not have it so you wouldn't be able to count on it. But I figure having it similar to optional `env.state()` would make sense.

**Comment by elliottower** [comment]
I don't have the bandwidth to do this any time soon but if a new contributor wants to help out with this it would be great
